[SPARK-35396]Add AutoCloseable close to BlockManager and InMemoryRelation

xuechendi opened the pull request:
This PR is proposing a add-on to support to manual close entries in MemoryStore and InMemoryRelation

### What changes were proposed in this pull request?
Currently:
    MemoryStore uses a LinkedHashMap[BlockId, MemoryEntry[_]] to store all OnHeap or OffHeap entries.
And when memoryStore.remove(blockId) is called, codes will simply remove one entry from LinkedHashMap and leverage Java GC to do release work.

This PR:
    We are proposing a add-on to manually close any object stored in MemoryStore and InMemoryRelation if this object is extended from AutoCloseable.

Veifiication:
    In our own use case, we implemented a user-defined off-heap-hashRelation for BHJ, and we verified that by adding this manual close, we can make sure our defined off-heap-hashRelation can be released when evict is called.
    Also, we implemented user-defined cachedBatch and will be release when InMemoryRelation.clearCache() is called by this PR 


### Why are the changes needed?
This changes can help to clean some off-heap user-defined object may be cached in InMemoryRelation or MemoryStore


### Does this PR introduce _any_ user-facing change?
NO


### How was this patch tested?
WIP

Signed-off-by: Chendi Xue <chendi.xue@intel.com>

AmplabJenkins commented:
Can one of the admins verify this patch?

srowen commented on core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala:
```diff
@@ -387,6 +388,14 @@ private[spark] class MemoryStore(
     val entry = entries.synchronized {
+      entries.get(blockId) match {
+        case e: DeserializedMemoryEntry[_] => e.value.foreach(v => v match {
+          case o: AutoCloseable =>
```
OK to this would cover plain Closeable too. I get the idea. On the one hand I don't think we _know_ it's safe to close, but also can't think of a case where you wouldn't, if it is closeable. Nothing else likely has a reference to it anyway.

The downside here is that you spend time closing, and we're in a synchronized block here, which is a little worrying. But it also may encounter an exception. Do you want to try-catch for NonFatal here so that it doesn't fail if close fails?

srowen replied:
I suppose you can at least only synchronize on entries to remove and get the remove elements; the rest can happen after the synchronized block. We're still holding the memoryManager lock here though.

xuechendi replied:
@srowen , I fixed the codes based on your last review, 1) directly using foreach { instead of foreach(v => v {  2) I added an async call here, so I hope close functions won't hold memoryManager lock for too long 3) Added a try catch to skip NonFatal exceptions.

mridulm replied:
In general, semantics of `AutoCloseable` has an expectation that it would be closed when released. I dont think we can make that assumption about `Closeable`.
I have not looked in greater detail if this assumption holds for Closeable for this usecase though.

xuechendi replied:
@mridulm , I see, to me, I think AutoCloseable is mostly commonly used, and close function implemented should be safe to close even more than once if it already be called by exception. Anyway, I can create a self-defined trait maybe named as "ManualCloseable", then any class extends from this "ManualCloseable" can be manual closed here, does that makes sense to you? @srowen , what do you think?

srowen replied:
I wouldn't make a new trait, no. I think anything AutoCloseable or Closeable is, well, closeable when done and should be closed at a point like this when we are sure (?) there are no other users of it.

xuechendi replied:
@srowen , thx, so ... Using AutoCloseable here as a hint to do manual close should be OK, am I understanding correctly?

srowen replied:
I think this is OK, as AutoCloseable is a supertype of Closeable, so this would handle both cases.

srowen commented on core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala:
```diff
@@ -387,6 +388,14 @@ private[spark] class MemoryStore(
 
   def remove(blockId: BlockId): Boolean = memoryManager.synchronized {
     val entry = entries.synchronized {
+      entries.get(blockId) match {
```
Doesn't remove already return what was removed? if so you can avoid .get - just .remove here and match on it

xuechendi replied:
@srowen , thx, I am a little concerned if I call entries.remove firstly, will the returned obj already be cleared and it may encounter some reference issue or other unexpected cases? so I still remained using get here.

srowen replied:
Why would the objects be gone? remove just removes copies of references from the map; they still exist and won't go away if you keep the references to them here.

srowen commented on core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala:
```diff
@@ -387,6 +388,14 @@ private[spark] class MemoryStore(
   def remove(blockId: BlockId): Boolean = memoryManager.synchronized {
     val entry = entries.synchronized {
+      entries.get(blockId) match {
+        case e: DeserializedMemoryEntry[_] => e.value.foreach(v => v match {
```
I think `.foreach(v => v match {` is redundant. Just `.foreach {`

srowen commented on core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala:
```diff
@@ -387,6 +388,14 @@ private[spark] class MemoryStore(
+        case e: DeserializedMemoryEntry[_] => e.value.foreach(v => v match {
+          case o: AutoCloseable =>
+              o.close
+            case other =>
```
Nit: indentation, and just write _, not other
Also put parens on .close; it has side effects.

xuechendi committed:
Add AutoCloseable close to MemoryStore

This commit is aimed to close some object may not be close by JVM GC
Fixes upon @srowen reviews:
1. Add try catch to skip NonFatal exception during close
2. Add an async caller to avoid close may hold BlockManager lock for too long
3. Add UT to clarify this PR

Signed-off-by: Chendi Xue <chendi.xue@intel.com>

xuechendi commented on core/src/test/scala/org/apache/spark/storage/MemoryStoreSuite.scala:
```diff
@@ -546,4 +551,105 @@ class MemoryStoreSuite
     }
   }
+
+  test("put native objects to MemoryStore and remove") {
```
@srowen , I added two UnitTests to verify this PR, my initial purpose of adding this Manual close is to release memory allocated by Apache-Arrow, so below UT is based on Arrow, while using Arrow in spark-core UT requires an extra dependencies in core's pom, I am not sure if it is ok to add dependencies in pom? Or I'll try to write something else to make the point.

srowen commented on core/src/test/scala/org/apache/spark/storage/MemoryStoreSuite.scala:
```diff
@@ -546,4 +550,105 @@ class MemoryStoreSuite
     }
   }
+
+  test("put native objects to MemoryStore and remove") {
```
This seems quite complex - is there any way to test this more easily? If it's this hard to test directly, I may say just punt on it. This is kind of best-effort to close the objects anyway.

xuechendi replied:
@srowen, I see, I changed the UT as a dummy one, using one allocated counter to verify close functions were called after remove() or clear()

srowen commented on core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala:
```diff
@@ -405,6 +427,23 @@ private[spark] class MemoryStore(
+      entries.values.asScala.foreach {
+        case e: DeserializedMemoryEntry[_] => e.value.foreach {
+          case o: AutoCloseable =>
+            entryManualCloseTasks += Future {
```
Not bad; I am a little concerned about the extra complexity and which thread pool this gets called in, maybe creating its own bottleneck. But you could argue it's better than waiting in clear(). Are any of the objects closed here expensive to close? I didn't check.

xuechendi replied:
@srowen , yes, your concern totally makes sense, now I changed to only use one Future to cleanup all, so cleanup won't start too much threads, and I added a 10ms wait in each close [UnitTest Link](https://github.com/apache/spark/pull/32534/files#diff-c4f6f7a030459e4a28c597761512bf4f02d4a213818da66525e0ba9c0327c47aR628-R637) to pretend as an expensive close.

xuechendi commented on core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala:
```diff
@@ -387,7 +392,24 @@ private[spark] class MemoryStore(
   def remove(blockId: BlockId): Boolean = memoryManager.synchronized {
     val entry = entries.synchronized {
-      entries.remove(blockId)
+      val removed = entries.remove(blockId)
```
@srowen , I see, I now changed it to use the object returned by remove.

srowen commented on core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala:
```diff
@@ -387,7 +392,24 @@ private[spark] class MemoryStore(
+          case _ =>
+        }
+      }
+      ThreadUtils.awaitResult(entryManualCloseTasks, Duration.Inf)
```
Do you need to await the results? it's best effort.

mridulm commented on core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala:
```diff
@@ -405,6 +427,22 @@ private[spark] class MemoryStore(
+          case _ =>
+        }
+      }
+      ThreadUtils.awaitResult(entryManualCloseTasks, Duration.Inf)
```
Pull this repeated code out into a method ?

xuechendi committed:
Fixed two issues upon review comments

1. use async onComplete to replace await.result
2. make duplicated manual_close to a function

Signed-off-by: Chendi Xue <chendi.xue@intel.com>

xuechendi commented on core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala:
```diff
@@ -385,9 +390,33 @@ private[spark] class MemoryStore(
+    entryManualCloseTasks.onComplete {
+      case Success(_) =>
+      case Failure(e) => throw e
+    }
```
@srowen , I replace ThreadUtils.awaitResult with onComplete here. And I also wrapped manual close as a function upon @mridulm 's suggestion.

xuechendi commented on core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala:
```diff
@@ -385,9 +390,33 @@ private[spark] class MemoryStore(
     val entry = entries.synchronized {
-      entries.remove(blockId)
+      val removed = entries.remove(blockId)
+      manual_close(removed)
```
@srowen @mridulm, manual_close is wrapped and called here.

xuechendi commented on core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala:
```diff
@@ -385,9 +390,33 @@ private[spark] class MemoryStore(
     }
   }
 
+  def manual_close[T <: MemoryEntry[_]](entry: T): T = {
```
@srowen, @mridulm , manual_close defined here.

srowen commented on core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala:
```diff
@@ -385,9 +390,33 @@ private[spark] class MemoryStore(
+        case _ =>
+      }
+    }
+    entryManualCloseTasks.onComplete {
```
This is still waiting on the tasks - I think we don't need to?

srowen replied:
OK I think you're right, it's not blocking

srowen replied:
This callback will throw an exception but not sure what that will do - probably just get logged. I think it's reasonable to log the error as a warning and continue - not being able to close is "OK" and shouldn't stop anything else

xuechendi replied:
@srowen , because NonFatal exception has been captured and logged in above try/catch, so I think only fail it will catch on this onComplete callback is Fatal exception, that is why I continue to throw here.

xuechendi replied:
@srowen , to make sure, I tested to throw one NonFatal and one Fatal errors in close function, and the behavior is as expected. 
```
case class NativeObject(alloc: DummyAllocator, size: Int)
      extends KnownSizeEstimation
      with AutoCloseable {

      alloc.alloc(size)
      var allocated_size: Int = size
      override def estimatedSize: Long = allocated_size
      override def close(): Unit = synchronized {
        Thread.sleep(10)
        alloc.release(allocated_size)
        allocated_size = 0
        throw new NotImplementedError()
        // throw new OutOfMemoryError()
      }
    }
```
![image](https://user-images.githubusercontent.com/4355494/119004235-f4fcce00-b9c0-11eb-9fbb-b6804efe6bd8.png)

![image](https://user-images.githubusercontent.com/4355494/119003362-3f317f80-b9c0-11eb-9c24-4c783bff3fe0.png)

and Then I tried to throw OutOfMemoryError as a FatalError, and UT will break
![image](https://user-images.githubusercontent.com/4355494/119004160-e57d8500-b9c0-11eb-9a08-bd4a3f2c61a9.png)

srowen commented on core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala:
```diff
@@ -385,9 +390,33 @@ private[spark] class MemoryStore(
     }
   }
 
+  def manual_close[T <: MemoryEntry[_]](entry: T): T = {
```
Nit: `manualClose`

xuechendi replied:
@srowen , fixed in latest commit, sorry for that

xuechendi committed:
Fix naming style issue and remove a empty for style check

Signed-off-by: Chendi Xue <chendi.xue@intel.com>

srowen commented:
Jenkins retest this please

SparkQA commented:
**[Test build #138835 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138835/testReport)** for PR 32534 at commit [`844f033`](https://github.com/apache/spark/commit/844f03385b3a713fd36ecd7c1bf9f783ac2b2949).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43357/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43357/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43357/

SparkQA commented:
**[Test build #138835 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138835/testReport)** for PR 32534 at commit [`844f033`](https://github.com/apache/spark/commit/844f03385b3a713fd36ecd7c1bf9f783ac2b2949).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/138835/

srowen commented:
Merged to master

cloud-fan commented on core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala:
```diff
@@ -21,9 +21,14 @@ import java.io.OutputStream
+import scala.collection.JavaConverters._
 import scala.collection.mutable
 import scala.collection.mutable.ArrayBuffer
+import scala.concurrent.ExecutionContext.Implicits.global
```
shall we create a dedicated thread pool (with a reasonable capacity) to run the closing tasks?

cloud-fan commented on core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala:
```diff
@@ -385,9 +390,33 @@ private[spark] class MemoryStore(
+              o.close
+            } catch {
+              case NonFatal(e) =>
+                logWarning(s"Got NonFatal exception during remove")
```
nit: `Fail to close a memory entry`

cloud-fan commented on core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala:
```diff
@@ -385,9 +390,33 @@ private[spark] class MemoryStore(
+    }
+    entryManualCloseTasks.onComplete {
+      case Success(_) =>
+      case Failure(e) => throw e
```
To confirm: this is a callback, so throwing an exception doesn't break the Spark application and just gets logged?

srowen replied:
From prior discussions, I believe that's right. In which case you could just log it
